File: backend/src/main/kotlin/taskmanager/backend/models/Work.kt
package taskmanager.backend.models

import org.bson.types.ObjectId
import taskmanager.backend.dtos.response.WorkResponseDto
import taskmanager.backend.models.base.impl.BaseEntityImpl
import taskmanager.backend.models.interfaces.AttachedToTaskEntity
import taskmanager.backend.models.interfaces.MappableCreatedByUserEntity
import java.util.Date

data class Work(
    override val createdBy: ObjectId,
    override val task: ObjectId,

    var description: String,
    var startDate: Date,
    var endDate: Date
) : BaseEntityImpl(), AttachedToTaskEntity, MappableCreatedByUserEntity<WorkResponseDto> {

    override fun toResponseDto(userId: ObjectId, createdBy: User?): WorkResponseDto {
        return WorkResponseDto(
            _id = _id.toString(),
            createdBy = createdBy?.toResponseDto(),
            task = task.toString(),
            description = description,
            startDate = startDate,
            endDate = endDate,
            createdAt = createdAt,
            updatedAt = updatedAt,
            canEdit = userId.toString() == userId.toString()
        )
    }
}

File: frontend/lib/models/work.dart
import 'package:frontend/models/base/base_entity.dart';
import 'package:frontend/models/user.dart';

class Work extends BaseEntity {
  User? createdBy;
  String task;
  String description;
  DateTime? startDate;
  DateTime? endDate;
  bool canEdit;

  Work({
    super.id,
    this.createdBy,
    this.task = '',
    this.description = '',
    this.startDate,
    this.endDate,
    super.createdAt,
    super.updatedAt,
    this.canEdit = false,
  });

  factory Work.fromJson(Map<String, dynamic> json) {
    return Work(
      id: json['_id'],
      createdBy: json['createdBy'] != null ? User.fromJson(json['createdBy'] as Map<String, dynamic>) : null,
      task: json['task'],
      description: json['description'],
      startDate: BaseEntity.parseDateFromJson(json['startDate']),
      endDate: BaseEntity.parseDateFromJson(json['endDate']),
      createdAt: BaseEntity.parseDateFromJson(json['createdAt']),
      updatedAt: BaseEntity.parseDateFromJson(json['updatedAt']),
      canEdit: json['canEdit'],
    );
  }

  String renderDates() {
    if (startDate == null || endDate == null) return '';

    String start = '${BaseEntity.renderDate(startDate!)} ${BaseEntity.renderTime(startDate!)}';
    String end = '${BaseEntity.renderDate(endDate!)} ${BaseEntity.renderTime(endDate!)}';

    return '$start - $end';
  }
}
